File: adt2play/txtscrio.pas
//  This file is part of Adlib Tracker II (AT2).
//
//  AT2 is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 3 of the License, or
//  (at your option) any later version.
//
//  AT2 is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with AT2.  If not, see <http://www.gnu.org/licenses/>.

unit TxtScrIO;
{$S-,Q-,R-,V-,B-,X+}
{$PACKRECORDS 1}
{$L txtscrio.o}
interface

{$I txtscrio/pas/colors.pas}

const
  MAX_SCREEN_MEM_SIZE = 180*60*2;
var
  SCREEN_MEM_SIZE: Longint; cvar; external;

type
  tSCREEN_MEM = array[0..PRED(MAX_SCREEN_MEM_SIZE)] of Byte;
  tSCREEN_MEM_PTR = ^tSCREEN_MEM;

var
  text_screen_shadow: tSCREEN_MEM; cvar; external;

  screen_ptr: Pointer; cvar; external;

  MaxCol: Byte; cvar; external;
  MaxLn:  Byte; cvar; external;

  cursor_backup: Longint; cvar; external;

procedure ShowC3Str(dest: tSCREEN_MEM_PTR; x,y: Byte; str: String; atr1,atr2,atr3: Byte); cdecl; external;

function  WhereX: Byte; cdecl; external;
function  WhereY: Byte; cdecl; external;
procedure GotoXY(x,y: Byte); cdecl; external;
function  GetCursor: Longint; cdecl; external;
procedure SetCursor(cursor: Longint); cdecl; external;
procedure ThinCursor; cdecl; external;
procedure WideCursor; cdecl; external;
procedure HideCursor; cdecl; external;
function  GetCursorShape: Word; cdecl; external;
procedure SetCursorShape(shape: Word); cdecl; external;

{$IFDEF GO32V2}

var
  v_seg:  Word; cvar; external;
  v_ofs:  Word; cvar; external;
  v_mode: Byte; cvar; external;

  DispPg: Byte; cvar; external;

function  iVGA: Boolean; cdecl; external;
procedure initialize; cdecl; external;
procedure GetRGBitem(color: Byte; var red,green,blue: Byte); cdecl; external;
procedure SetRGBitem(color: Byte; red,green,blue: Byte); cdecl; external;
procedure GetPalette(var pal; first,last: Word); cdecl; external;
procedure SetPalette(var pal; first,last: Word); cdecl; external;
procedure WaitRetrace; cdecl; external;

{$I txtscrio/go32/pas/fade.pas}

procedure VgaFade(var data: tFADE_BUF; fade: tFADE; delay: tDELAY); cdecl; external;

type
  tVIDEO_STATE = Record
                   cursor: Longint;
                   font: Byte;
                   MaxLn,MaxCol,v_mode,DispPg: Byte;
                   v_seg,v_ofs: Word;
                   screen: tSCREEN_MEM;
                   data: array[0..PRED(4096)] of Byte;
                 end;

procedure GetVideoState(var data: tVIDEO_STATE); cdecl; external;
procedure SetVideoState(var data: tVIDEO_STATE; restore_screen: Boolean); cdecl; external;

{$ENDIF}

implementation

uses
{$IFDEF GO32V2}
  GO32,
{$ENDIF}
  pascal,
  common;

begin
  initialize;
end.

File: adt2play/txtscrio.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

#include "defines.h"

#include <stddef.h>
#if USE_FPC
#include "pascal.h"
#if GO32
#include "pascal/farptr.h"
#include "pascal/go32.h"
#include "pascal/pc.h"
#endif // GO32
#include "pascal/dos.h"
#include "pascal/string.h"
#else // !USE_FPC
#if GO32
#include <farptr.h>
#include <go32.h>
#include <pc.h>
#endif // GO32
#include <dos.h>
#include <string.h>
#endif // !USE_FPC
#if GO32
#include "go32/VBIOS.h"
#include "go32/VGA.h"
#endif // GO32
#include "common.h"
#include "txtscrio.h"

// HINT: (FPC) S-: Stack checking (off)
// HINT: (FPC) Q-: Overflow checking (off)
// HINT: (FPC) R-: Range checking (off)
// HINT: (FPC) V-: Var-string checking (off)
// HINT: (FPC) B-: Complete boolean evaluation (off)
// HINT: (FPC) X+: Extended syntax (ON)
// HINT: (FPC) PACKRECORDS 1: Alignment of record elements (1)

int32_t SCREEN_MEM_SIZE = MAX_SCREEN_MEM_SIZE;

tSCREEN_MEM text_screen_shadow;

void *screen_ptr = &text_screen_shadow;

uint8_t MaxCol = 80;
uint8_t MaxLn = 25;

int32_t cursor_backup;

static uint16_t absolute_pos;

#include "txtscrio/PosChar.c"

#include "txtscrio/ShowC3Str.c"

#include "txtscrio/WhereX.c"
#include "txtscrio/WhereY.c"
#include "txtscrio/GotoXY.c"
#include "txtscrio/GetCursor.c"
#include "txtscrio/SetCursor.c"
#include "txtscrio/ThinCursor.c"
#include "txtscrio/WideCursor.c"
#include "txtscrio/HideCursor.c"
#include "txtscrio/GetCursorShape.c"
#include "txtscrio/SetCursorShape.c"

uint16_t v_seg = 0xB800;
uint16_t v_ofs = 0;
uint8_t  v_mode = 3;

uint8_t DispPg;

#include "txtscrio/go32/iVGA.c"
#include "txtscrio/go32/initialize.c"
#include "txtscrio/go32/GetRGBitem.c"
#include "txtscrio/go32/SetRGBitem.c"
#include "txtscrio/go32/GetPalette.c"
#include "txtscrio/go32/SetPalette.c"
#include "txtscrio/go32/WaitRetrace.c"

#include "txtscrio/go32/fade.c"

#include "txtscrio/go32/VgaFade.c"
